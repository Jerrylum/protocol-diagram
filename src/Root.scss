@use "style/Essential.scss";
@use "style/Modals.scss";

#root {
  overflow: hidden;

  --bg-body-color: #ffffff;
  --bg-default-color: #ffffff;
  --bg-paper-color: #f3f3f3;
  --text-primary-color: rgba(0, 0, 0, 0.87);
  --text-disabled-color: rgba(0, 0, 0, 0.6);
  --link-color: rgba(0, 0, 0, 0.87);
  --hover-color: rgba(0, 0, 0, 0.08);
  --focused-color: rgba(0, 0, 0, 0.12);
  background-color: var(--bg-body-color);

  > #root-container {
    display: flex;
    width: 100vw;
    width: 100svw;
    height: 100vh;
    height: 100svh;
    box-sizing: border-box;
  }
}
