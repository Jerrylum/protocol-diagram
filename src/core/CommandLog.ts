import { makeAutoObservable } from "mobx";

export type LogLevel = "info" | "error";

export interface CommandLog {
  level: LogLevel;
  message: string;
}

export class CommandLogger {
  private _logs: CommandLog[] = [];
  private _logCount = 0;

  constructor(public _maxLogs: number = 100) {
    makeAutoObservable(this);
  }

  get logs(): CommandLog[] {
    return this._logs;
  }

  get logCount(): number {
    return this._logCount;
  }

  add(level: LogLevel, message: string) {
    this._logs.push({ level, message });
    this._logs = this._logs.slice(-this._maxLogs);
    this._logCount++;
  }

  info(message: string) {
    this.add("info", message);
  }

  error(message: string) {
    this.add("error", message);
  }

  clear() {
    this._logs = [];
  }

  get maxLogs(): number {
    return this._maxLogs;
  }

  set maxLogs(maxLogs: number) {
    this._maxLogs = maxLogs;
    this._logs = this._logs.slice(-maxLogs);
  }
}
