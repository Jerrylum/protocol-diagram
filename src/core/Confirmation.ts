import { makeAutoObservable, when } from "mobx";
import { Modals } from "./Modals";

export interface ConfirmationButton {
  label: string;
  onClick?: () => void;
  hotkey?: string;
  color?: "inherit" | "primary" | "secondary" | "success" | "error" | "info" | "warning";
}

export interface ConfirmationPromptData {
  title: string;
  description: React.ReactNode;
  buttons: ConfirmationButton[];
  onKeyDown?(e: React.KeyboardEvent<HTMLDivElement>, onClick: (index: number) => void): void;
}

export interface ConfirmationInputPromptData extends ConfirmationPromptData {
  inputLabel: string;
  inputDefaultValue: string;
}

export class Confirmation {
  private data?: ConfirmationPromptData | ConfirmationInputPromptData;
  public input?: string;

  constructor(private modals: () => Modals) {
    makeAutoObservable(this);
  }

  close() {
    this.modals().close(ConfirmationModalSymbol);
    this.data = undefined;
  }

  async prompt(data: ConfirmationPromptData | ConfirmationInputPromptData, priority: number = 1) {
    if (data.buttons.length === 0) {
      data.buttons.push({ label: "OK" });
    }

    this.close();
    this.data = data;
    if ("inputLabel" in data && "inputDefaultValue" in data) {
      this.input = data.inputDefaultValue;
    } else {
      this.input = undefined;
    }

    this.modals().open(ConfirmationModalSymbol, priority);

    await when(() => this.data === undefined);

    return this.input;
  }

  get isOpen() {
    return this.data !== undefined && this.modals().opening === ConfirmationModalSymbol;
  }

  get title() {
    return this.data?.title ?? "";
  }

  set title(value: string) {
    if (this.data === undefined) return;
    this.data.title = value;
  }

  get description() {
    return this.data?.description ?? "";
  }

  set description(value: React.ReactNode) {
    if (this.data === undefined) return;
    this.data.description = value;
  }

  get buttons() {
    return this.data?.buttons ?? [];
  }

  get onKeyDown() {
    return this.data?.onKeyDown;
  }

  get inputLabel() {
    return (this.data as ConfirmationInputPromptData)?.inputLabel ?? "";
  }
}

export const ConfirmationModalSymbol = Symbol("ConfirmationModalSymbol");
