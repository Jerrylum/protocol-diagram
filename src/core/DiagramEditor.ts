import { makeAutoObservable } from "mobx";
import { clamp } from "./Util";
import { Vector } from "./Vector";

export class DiagramEditor {
  private _offset: Vector = new Vector(0, 0);
  private _scale: number = 1; // 1 = 100%, [1..3]

  get offset() {
    return this._offset;
  }

  get scale() {
    return this._scale;
  }

  set offset(offset: Vector) {
    this._offset = offset;
  }

  set scale(scale: number) {
    this._scale = clamp(scale, 0.75, 2);
  }

  constructor() {
    makeAutoObservable(this);
  }

  resetOffsetAndScale() {
    this.offset = new Vector(0, 0);
    this.scale = 1;
  }
}