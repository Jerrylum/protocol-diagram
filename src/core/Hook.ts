import React from "react";

export function useBetterMemo<T extends { destructor: () => void } | {}>(
  factory: () => T,
  deps: React.DependencyList = []
): T {
  const storage = React.useMemo(() => factory(), deps); // eslint-disable-line react-hooks/exhaustive-deps
  React.useEffect(() => {
    return () => {
      if ("destructor" in storage && typeof storage.destructor === "function") storage.destructor?.();
    };
  }, deps); // eslint-disable-line react-hooks/exhaustive-deps

  return storage;
}
