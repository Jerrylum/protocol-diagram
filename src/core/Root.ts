import { Modals } from "./Modals";

export type RootStore = typeof rootStore;

const rootStore = {
  modals: new Modals()
} as const;

export function getRootStore(): RootStore {
  return rootStore;
}

