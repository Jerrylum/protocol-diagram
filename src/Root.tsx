import "./Root.scss";
import { observer } from "mobx-react-lite";
import { DiagramCanvas } from "./app/DiagramCanvas";
import { Box } from "@mui/material";
import { BottomPanel } from "./app/BottomPanel";
// import { TestModal, TestModalSymbol } from "./app/TestModal";
// import React from "react";
// import { getRootStore } from "./core/Root";

const Root = observer(() => {
  // React.useEffect(() => {
  //   getRootStore().modals.open(TestModalSymbol);
  // }, []);
  return (
    <Box id="root-container">
      <DiagramCanvas />
      <BottomPanel />
      {/* <TestModal /> */}
    </Box>
  );
});

export default Root;

