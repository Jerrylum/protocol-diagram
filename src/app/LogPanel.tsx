import { Box, Typography } from "@mui/material";
import { observer } from "mobx-react-lite";
import React from "react";
import { useTimeout } from "../core/Hook";
import { CommandLog } from "../core/CommandLog";
import { getRootStore } from "../core/Root";

export const CommandLogComponent = observer((props: { log: CommandLog; focused: boolean }) => {
  const log = props.log;

  const [hiding, setHiding] = React.useState(false);

  React.useEffect(() => {
    if (props.focused) {
      setHiding(false);
    }
  }, [props.focused]);

  useTimeout(
    () => {
      setHiding(true);
    },
    props.focused ? null : 2000,
    [props.focused]
  );

  return (
    <Typography
      variant="body2"
      sx={{
        fontFamily: "Ubuntu Mono",
        color: log.level === "info" ? "text.primary" : "error.main",
        backgroundColor: "rgba(255, 255, 255, 0.6)",
        opacity: hiding ? 0 : 1,
        transition: hiding ? "opacity 2s ease-in-out" : "",
        minHeight: "20px",
        lineHeight: "20px",
        padding: "0 4px"
      }}>
      {log.message}
    </Typography>
  );
});

export const LogPanel = observer(() => {
  const { logger } = getRootStore();
  const logs = logger.logs;

  const [focused, setFocused] = React.useState(false);
  const logPanelRef = React.useRef<HTMLDivElement>(null);

  React.useEffect(() => {
    if (logPanelRef.current) {
      logPanelRef.current.scrollTop = logPanelRef.current.scrollHeight;
    }
  }, [logger.logCount]);

  return (
    <Box
      ref={logPanelRef}
      onMouseEnter={() => setFocused(true)}
      onMouseLeave={() => setFocused(false)}
      sx={{
        position: "absolute",
        left: "0",
        right: "0",
        bottom: "calc(100% + 4px)",
        maxHeight: "60px",
        overflowY: "auto"
      }}>
      {logs.map((log, i) => (
        <CommandLogComponent key={log.uid} log={log} focused={focused} />
      ))}
    </Box>
  );
});
