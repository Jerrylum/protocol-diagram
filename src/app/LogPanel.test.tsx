import { fireEvent, render } from "@testing-library/react";
import { LogPanel } from "./LogPanel";
import { act } from "react-dom/test-utils";
import { getRootStore } from "../core/Root";

test("Render LogPanel", () => {
  const result = render(<LogPanel />);
  expect(result.container).toMatchSnapshot();
});

test("Render LogPanel with focused", () => {
  jest.useFakeTimers();
  const { logger } = getRootStore();
  logger.add("error", "test");
  const components = (
    <div id="root-container">
      <LogPanel />
    </div>
  );
  const result = render(components);
  const logPanel = result.container.querySelector("#root-container > div");
  act(() => {
    fireEvent.mouseEnter(logPanel!);
  });

  act(() => {
    fireEvent.mouseLeave(logPanel!);
  });
  result.rerender(components);
  jest.runAllTimers();
  result.rerender(components);
});
