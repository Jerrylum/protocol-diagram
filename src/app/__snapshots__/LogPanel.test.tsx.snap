// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render LogPanel 1`] = `
<div>
  <div
    class="MuiBox-root css-1e4trel"
  />
</div>
`;
